central/reports/service/v2/service_impl.go

```diff
@@ -115,3 +122,60 @@ func (s *serviceImpl) GetReportHistory(ctx context.Context, req *apiV2.GetReport
 	}
 	return &res, nil
 }
+
+func (s *serviceImpl) RunReport(ctx context.Context, req *apiV2.RunReportRequest) (*apiV2.RunReportResponse, error) {
+	if req.GetReportConfigId() == "" {
+		return nil, errors.Wrap(errox.InvalidArgs, "Report configuration ID is empty")
+	}
+	slimUser := authn.UserFromContext(ctx)
+	if slimUser == nil {
+		return nil, errors.New("Could not determine user identity from provided context")
+	}
+	config, found, err := s.reportConfigStore.GetReportConfiguration(ctx, req.GetReportConfigId())
+	if err != nil {
+		return nil, errors.Wrapf(err, "Error finding report configuration %s", req.GetReportConfigId())
+	}
+	if !found {
+		return nil, errors.Wrapf(errox.NotFound, "Report configuration id not found %s", req.GetReportConfigId())
+	}
+	reportReq := &reportGen.ReportRequest{
+		ReportConfig: config,
+		ReportMetadata: &storage.ReportMetadata{
+			ReportConfigId: req.GetReportConfigId(),
+			Requester:      slimUser,
+			ReportStatus: &storage.ReportStatus{
+				RunState:          storage.ReportStatus_WAITING,
+				ReportRequestType: storage.ReportStatus_ON_DEMAND,
+			},
+		},
+	}
+	if req.GetReportNotificationMethod() == apiV2.NotificationMethod_EMAIL {
+		reportReq.ReportMetadata.ReportStatus.ReportNotificationMethod = storage.ReportStatus_EMAIL
+	} else {
+		reportReq.ReportMetadata.ReportStatus.ReportNotificationMethod = storage.ReportStatus_DOWNLOAD
+		// Scope the downloadable reports to the access scope of user demanding the report
+		reportReq.Ctx = ctx
+	}
+	reportID, err := s.scheduler.SubmitReportRequest(reportReq, false)
```

Comment:
Yeah, that check happens in the implementation of `SubmitReportRequest` . The reason I am not able to do it here at service layer is because I need to do both these things inside a lock as a single transaction: 1) check if another job is in WAITING or PREPARING and 2) store this job to database in WAITING state. 

With the current implementation, doing those two things at separate places could cause a race conditions. If a user accidentally double clicks "Run report" both jobs can get submitted. So in order to do those two things together, I added the WAITING or PREPARING check inside `scheduler.SubmitRequest` too.